router: Allow using another header for virtual host matching

pugmajere opened the pull request:
This adds a new RouteConfiguration option that specifies an alternative header to be used instead of :authority/Host when selecting a virtual host

This allows for a "simplification" or "canonicalization" step to be run on :authority, to reduce high cardinality portions of the hostname to be replaced with a static string, without impacting that :authority header that is sent upstream. This provides a way to support multiple wildcards in a URL in a clean way.

Commit Message: router: Allow using another header for virtual host matching
Additional Description:
This adds a new RouteConfiguration option that specifies an alternative header to be used instead of :authority/Host when selecting a virtual host

This allows for a "simplification" or "canonicalization" step to be run on :authority, to reduce high cardinality portions of the hostname to be replaced with a static string, without impacting that :authority header that is sent upstream. This provides a way to support multiple wildcards in a URL in a clean way.

Risk Level:
Testing: added new tests
Docs Changes: Updated api definition
Release Notes: Updated changelog
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

wbpcode commented on api/envoy/config/route/v3/route.proto:
```diff
@@ -133,6 +133,13 @@ message RouteConfiguration {
   // :ref:`envoy_v3_api_msg_config.route.v3.VirtualHost`.domains field.
   bool ignore_port_in_host_matching = 14;
 
+  // Normally, virtual host matching is done using the :authority (or
+  // Host: in HTTP < 2) HTTP header. Setting this will instead, use a
+  // different HTTP header for this purpose. This is intended to be
+  // combined with an "early_header_mutation" extension to allow
+  // alternate or simplified host values to be used for host matching.
+  string alternate_header_for_host_matching = 18;
```
`vhost_header` is enough. Because most users could get it is header name that will be used for vhost searching from the simple `vhost_header`.

The comments also could be simplified and needn't to refer the early header mutation. The new feature self should be a general/common feature and needn't be bound to early header mutation

wbpcode reviewed (changes requested):
Thanks for the contribution. And sorry for the delayed response. I am still OOO and cannot review code in time.

pugmajere commented on api/envoy/config/common/mutation_rules/v3/mutation_rules.proto:
```diff
@@ -97,6 +97,16 @@ message HeaderMutation {
+
+    type.matcher.v3.RegexMatchAndSubstitute expression = 3;
+  }
+
```
Ah, the CEL support is ... not well documented (I was aware of the library, but I've never seen anything that provided an example like this before, well, this.)

Removed this early header mutation extension update as a result.

wbpcode replied:
Yeah. TBH, I think only rare users need this because it's too flexible.

pugmajere commented:
I've updated and rebased, but left things as 5 commits so you can see the changes. Also, included a changelog update now. (That kinda forced the rebase, tbh.)

wbpcode commented on source/common/router/config_impl.h:
```diff
@@ -1234,6 +1234,7 @@ class RouteMatcher {
 
   VirtualHostImplSharedPtr default_virtual_host_;
   const bool ignore_port_in_host_matching_{false};
+  const std::string vhost_header_;
```
```suggestion
  const Http::LowerCaseString vhost_header_;
```

to avoid create the lower case string copy for every match.

wbpcode commented:
Seems main merge is necessary.

wbpcode commented:
/retest

wbpcode reviewed (approved):
LGTM. Thanks.

tyxia reviewed (approved):
wbpcode has approved it.

Thank you for your contribution!
